stage0/src/sev.rs

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use oak_linux_boot_params::{BootParams, E820EntryType};
+use sev_guest::instructions::{pvalidate, InstructionError, PageSize as SevPageSize, Validation};
+use x86_64::{
+    align_down, align_up,
+    instructions::tlb,
+    registers::control::Cr3,
+    structures::paging::{PageSize, PageTable, PageTableFlags, PhysFrame, Size2MiB, Size4KiB},
+    PhysAddr,
+};
+
+/// Calls `PVALIDATE` on all memory ranges specified in the E820 table with type `RAM`.
+pub fn validate_memory(zero_page: &BootParams, encrypted: u64) {
+    let mut page_table = PageTable::new();
+
+    // Find a location for our (temporary) page table. The initial page tables map [0..2MiB), so it
+    // should be safe to put our page table to [2..4MiB).
```

Comment:
nit: I found this comment a bit confusing on the first pass through, as the page table itself is not located in this range - it just map values in this range to physical addresses.

Reply:
Done.